First draft of the Graph guidelines (#303)
Microsoft/api-guidelines: graph/GuidelinesGraph.md

# Microsoft Graph REST API Guidelines

Table of Contents

- [Microsoft Graph REST API Guidelines](#microsoft-graph-rest-api-guidelines)
  - [](#)
      - [History](#history)
  - [Introduction](#introduction)
    - [Legend](#legend)
  - [Design Approach](#design-approach)
    - [Naming](#naming)
    - [Uniform Resource Locators (URLs)](#uniform-resource-locators-urls)
    - [Query Support](#query-support)
    - [Resource Modeling Patterns](#resource-modeling-patterns)
    - [Behavior Modeling](#behavior-modeling)
    - [Error Handling](#error-handling)
  - [API contract and non-backward compatible changes](#api-contract-and-non-backward-compatible-changes)
    - [Versioning and Deprecation](#versioning-and-deprecation)
  - [Recommended API Patterns](#recommended-api-patterns)
  - [References](#references)

## 

#### History

| Date        | Notes                       |
|-------------|-----------------------------|
| 2021-Sep-28 | Using summary and patterns style. |
| 2020-Oct-04 | Initial version in Wiki.    |

## Introduction

> When building a digital ecosystem API usability becomes a business priority. Success of your ecosystem depends on APIs that are easy to discover, simple to use, fit to purpose, and consistent across your products.

Reviewer: ```suggestion
When building a digital ecosystem API usability becomes a business priority. Success of your ecosystem depends on APIs that are easy to discover, simple to use, fit for purpose, and consistent across your products.
```

> This document offers guidance that Microsoft Graph API developer teams MUST follow to

Reviewer: ```suggestion
This document offers guidance that Microsoft Graph API producer teams MUST follow to
```

ensure that Microsoft Graph has a consistent and easy to use API surface. A new API design should meet the
following goals:

\- Developer friendly via consistent naming, patterns, and web standards (HTTP,
REST, JSON)

\- Work well with SDKs in many programming languages.

\- Sustainable & evolvable via clear API contracts.

The Microsoft Graph guidelines are an extension of the Microsoft REST API
guidelines. Readers are assumed also be reading and following the Microsoft REST API
guidelines except where this document outlines specific differences or exceptions to those guidelines.
Together these guidelines and a library of API patterns serve as the means by

> which API teams discuss and come to consensus on API review recommendations.

Reviewer: ```suggestion
which API teams discuss and come to consensus on API review requirements..
```

This document borrows from multiple public sources such as:

1.  Microsoft Azure REST API Guidelines

2.  Google Cloud Platform APIs

3.  WSO2 Rest API Design Guidelines and others.

Technology and software are constantly changing and evolving, and as such, this
is intended to be a living document. [Open an
issue](https://github.com/microsoft/api-guidelines/issues/new/choose) to suggest
a change or propose a new idea.

> 

Reviewer: API guidelines that change frequently lead to an uneven and inconsistent API surface.  Consequently, this document will more frequently change to add guidance in areas previously uncovered, or to clarify existing guidance.   It will less frequently change the directional guidance it has already provided.

### Legend

This document offers prescriptive guidance labeled as follows:

:heavy_check_mark: **MUST** satisfy this specification. If not following this
advice, you MUST disclose your reason during the Graph API review.

:no_entry: **MUST NOT** use this pattern. If not following this advice, you MUST
disclose your reason during the Graph API review.

:ballot_box_with_check: **SHOULD** fulfill this specification. If not
following this advice, you MUST disclose your reason during the Graph API
review.

> :warning: **SHOULD NOT** adopt this pattern. If not following this advice, you MUST disclose your reason during the Graph API review.

Reviewer: Don't repeat this "If not following sentence"  Put it at the top or bottom instead.

## Design Approach

> The design of your API is arguably the most important investment you will make. API design is what creates the first impression for developers when they discover and learn how to use your APIs. We promote API-first design approach where you begin your product design by focusing on how information will be exchanged and represented and creating an interface contract for your API which is followed by design and implementation of the backing service. This approach ensures decoupling of the interface and is essential for agility, predictability, and reuse of your APIs. Starting with user-facing contracts also promotes good understanding of user interactions, your modeling domain, and understanding of how the service will evolve. Microsoft Graph supports resource and query based API styles which follow HTTP, REST, and JSON standards, where API contract is described using ODATA conventions and schema definition (see Documentation · OData - the Best Way to REST).

Reviewer: ```suggestion
The design of your API is arguably the most important and long-term investment you will make. API design is what creates the first impression for developers when they discover and learn how to use your APIs. We promote API-first design approach where you begin your product design by focusing on how information will be exchanged and represented and creating an interface contract for your API which is followed by design and implementation of the backing service. This approach ensures decoupling of the interface and is essential for agility, predictability, and reuse of your APIs. Starting with user-facing contracts also promotes good understanding of user interactions, your modeling domain, and understanding of how the service will evolve. Microsoft Graph supports resource and query based API styles which follow HTTP, REST, and JSON standards, where API contract is described using ODATA conventions and schema definition (see Documentation · OData - the Best Way to REST).
```

  - [](#)
      - [History](#history)
  - [Introduction](#introduction)
    - [Legend](#legend)
  - [Design Approach](#design-approach)
    - [Naming](#naming)
    - [Uniform Resource Locators (URLs)](#uniform-resource-locators-urls)
    - [Query Support](#query-support)
    - [Resource Modeling Patterns](#resource-modeling-patterns)
    - [Behavior Modeling](#behavior-modeling)
    - [Error Handling](#error-handling)
  - [API contract and non-backward compatible changes](#api-contract-and-non-backward-compatible-changes)
    - [Versioning and Deprecation](#versioning-and-deprecation)
  - [Recommended API Patterns](#recommended-api-patterns)
  - [References](#references)

## 

#### History

| Date        | Notes                       |
|-------------|-----------------------------|
| 2021-Sep-28 | Using summary and patterns style. |
| 2020-Oct-04 | Initial version in Wiki.    |

## Introduction

When building a digital ecosystem API usability becomes a business priority. Success of your ecosystem depends on APIs that are easy to discover, simple to use, fit to purpose, and consistent across your products.

> This document offers guidance that Graph API developer teams MUST follow to

Reviewer 2: ```suggestion
This document offers guidance that Microsoft Graph API developer teams MUST follow to
```

ensure that Microsoft Graph has a consistent and easy to use API surface. A new API design should meet the
following goals:

\- Developer friendly via consistent naming, patterns, and web standards (HTTP,
REST, JSON)

\- Work well with SDKs in many programming languages.

\- Sustainable & evolvable via clear API contracts.

The Microsoft Graph guidelines are an extension of the Microsoft REST API
guidelines. Readers are assumed also be reading and following the Microsoft REST API
guidelines except where this document outlines specific differences or exceptions to those guidelines.
Together these guidelines and a library of API patterns serve as the means by
which API teams discuss and come to consensus on API review recommendations.

This document borrows from multiple public sources such as:

1.  Microsoft Azure REST API Guidelines

2.  Google Cloud Platform APIs

> 3.  WSO2 Rest API Design Guidelines and others.

Reviewer 2: We do?

Author: we do in spirit but it also the old version of the guidelines

Technology and software are constantly changing and evolving, and as such, this
is intended to be a living document. [Open an
issue](https://github.com/microsoft/api-guidelines/issues/new/choose) to suggest
a change or propose a new idea.

### Legend

This document offers prescriptive guidance labeled as follows:

> :heavy_check_mark: **MUST** satisfy this specification. If not following this

Reviewer 2: ```suggestion
:heavy_check_mark: **MUST** satisfy this guidance. If not following this
```
?

## Design Approach

The design of your API is arguably the most important investment you will make. API design is what creates the first impression for developers when they discover and learn how to use your APIs. We promote API-first design approach where you begin your product design by focusing on how information will be exchanged and represented and creating an interface contract for your API which is followed by design and implementation of the backing service. This approach ensures decoupling of the interface and is essential for agility, predictability, and reuse of your APIs. Starting with user-facing contracts also promotes good understanding of user interactions, your modeling domain, and understanding of how the service will evolve. Microsoft Graph supports resource and query based API styles which follow HTTP, REST, and JSON standards, where API contract is described using ODATA conventions and schema definition (see Documentation · OData - the Best Way to REST).
[Documentation · OData - the Best Way to REST](https://www.odata.org/documentation/)).

In general API design includes the following steps:

-   Define your domain model

-   Derive and name your API resources

> 

Reviewer 2: + Describe relationships between resources

-   Specify errors

When creating your API contract you will define resources based on the domain model supporting your service and identify interactions based on use cases. Good API design goes beyond modeling the current state of resources and it is important to plan ahead how API evolves. For this it is essential to understand and document your use cases as the foundation of the API design. There is no one-to-one correspondence between domain model elements and API resources because you should simplify your customer facing APIs for better usability and to obfuscate implementation details. We recommend creating a simple resource diagram, like below, to show resources and their relationships and make it easier to reason about modeling choices and the shape of your API.
![Resource model example](ModelExample.png)

After resources are defined it’s time to think about the behavior of your API which can be expressed via HTTP methods and operational resources such as functions and actions. As you think about API behavior you identify a happy path and various exceptions and deviations which will be expressed as errors and represented using HTTP codes and error messages.

> At every step of your design you need to consider security, privacy and compliance as an intrinsic components of your API implementation.

Reviewer 2: ```suggestion
At every step of your design you need to consider security, privacy and compliance as intrinsic components of your API implementation.
```

    Microsoft products, i.e. /search, /notifications, /subscriptions, /files.

4.  *Administrative configuration* functions for specific products. (Note: this
    is not final and may be adjusted based on the survey results), i.e. /admin/exchange.

5.  Internal Microsoft requirements for publishing Privileged and Hidden APIs,
    routing, and load testing, i.e./loadTestEntities.

Effectively top-level categories define a perimeter for the API surface thus a
new category creation requires additional rigor and governance.

### Query Support

Microsoft Graph APIs should support basic query options in conformance with
OData specifications and [Microsoft REST API
Guidelines](https://github.com/microsoft/api-guidelines/blob/master/Guidelines.md#7102-error-condition-responses).
|Requirements|
|----------------------------------------------------------------------------------------------------|
| :ballot_box_with_check: **SHOULD** support \$select on resource to enable properties projection |
|:ballot_box_with_check: **SHOULD** support \$filter with eq, ne operations on properties of entities for collections| :heavy_check_mark:

> | :ballot_box_with_check: **SHOULD** support pagination 4top and $count for collections |

Reviewer 2: ```suggestion
| :ballot_box_with_check: **SHOULD** support pagination $top and $count for collections |
```

Limitations of \$query requests made to Microsoft Graph:

-   Microsoft Graph only supports having all the query options completely in the
    request body or completely in the request url. Graph doesn't support query
    options present in both places.

-   The parameters in \$query should not span multiple workloads. Support for
    \$query right now is limited to properties belonging to the same workload.

The query options part of an OData URL can be quite long, potentially exceeding
the maximum length of URLs supported by components involved in transmitting or
processing the request. One way to avoid this is to use the POST verb instead of
GET, and pass the query options part of the URL in the request body as described
in the chapter [OData Query
Options](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part2-url-conventions.html#sec_PassingQueryOptionsintheRequestBody).

### Resource Modeling Patterns

> You can model complex resources for your APIs using OData Entity Type or Complex Type. The main difference between these types is that Entity types declare a key property to uniquely identify its objects and  Complex Types don't. In Microsoft Graph this key property has "id" as a prescribed name.

Reviewer 2: ```suggestion
You can model complex resources for your APIs using an OData Entity Type or Complex Type. The main difference between these types is that Entity types declare a key property to uniquely identify its objects and Complex types do not. In Microsoft Graph this key property has "id" as a prescribed name.
```

> Since objects of complex types on Graph don’t have unique identifiers, they are not directly addressable via URIs and therefore you must use Entity Type to model addressable resources. Complex types are better suited to represent composite properties of API entities.

Reviewer 2: ```suggestion
Since objects of complex types on Graph SHOULD not have unique identifiers, they are not directly addressable via URIs and therefore you MUST use Entity Type to model addressable resources. Complex types are better suited to represent composite properties of API entities.
```

| :heavy_check_mark: **MUST** return a 429 Too many requests error when the caller has exceeded throttling limits.        | 
| :ballot_box_with_check: **SHOULD** return a 404 Not found error if a 403 would result in information disclosure. |

For a complete mapping of error codes to HTTP statuses you can refer to the
[rfc7231 (ietf.org)](https://datatracker.ietf.org/doc/html/rfc7231#section-6).

## API contract and non-backward compatible changes

Microsoft Graph definition of breaking changes is based on the [Microsoft REST
API
Guidelines](https://github.com/microsoft/api-guidelines/blob/graph/Guidelines.md#123-definition-of-a-breaking-change).
In general, making changes to the API contract for existing elements is
considered breaking. Adding new elements is allowed and not considered a
breaking change.

\*\* Non-breaking changes:\*\*

-   Addition of an annotation OpenType="true" 
-   Addition of properties that are nullable or have a default value
-   Addition of a member to an evolvable enumeration 
-   Removal, rename, or change to the type of an open extension
-   Removal, rename, or change to the type of an annotation 

> -   Introduction of paging to existing collections

Reviewer 2: We consider introduction of paging a breaking change

-   **Changes to error codes?????**
-   Changes to the order of properties
-   Changes to the length or format of opaque strings, such as resource IDs

\*\* Breaking changes:\*\*

-   Changes to the URL or fundamental request/response associated with a
    resource
-   Changing semantics of resource representation

> -   Removal, rename, or change to the type of a declared property

Reviewer 2: ```suggestion
-   Removal, rename, or change to an incompatible type of a declared property
```

-   Removal or rename of APIs or API parameters
-   Addition of a required request header
-   Addition of a EnumType members for non-extensible enumerations
-   Addition of a Nullable="false" properties to existing types
-   Addition of a Nullable="false" parameters to existing actions and functions

> -   **Adding attributes to existing nodes is considered breaking??**

Reviewer 2: I don't know what this means. I would remove it.

For the full list of rules you can refer to [this section of the OData V4

> spec](https://docs.oasis-open.org/odata/odata/v4.0/errata02/os/complete/part1-protocol/odata-v4.0-errata02-os-part1-protocol-complete.html#_Toc406398209).

Reviewer 2: We should not do this.  OData does not define the compatibility rules for Microsoft Graph.

Author: ok will remove

      - [History](#history)
  - [Introduction](#introduction)
    - [Legend](#legend)
  - [Design Approach](#design-approach)
    - [Naming](#naming)
      - [Casing](#casing)
      - [Prefixes and Suffixes](#prefixes-and-suffixes)
    - [Uniform Resource Locators (URLs)](#uniform-resource-locators-urls)
      - [Category](#category)
    - [Query Support](#query-support)
    - [Recommended Modeling Patterns](#recommended-modeling-patterns)
  - [Behavior Modeling](#behavior-modeling)
      - [HTTP Operations](#http-operations)
      - [Microsoft Graph rules for modeling behavior](#microsoft-graph-rules-for-modeling-behavior)
    - [Error Handling](#error-handling)
    - [API contract and non-backward compatible changes](#api-contract-and-non-backward-compatible-changes)
  - [Versioning and Deprecation](#versioning-and-deprecation)
    - [Deprecation Process](#deprecation-process)
  - [Common API Patterns](#common-api-patterns)
  - [References](#references)

## 
#### History

| Date        | Notes                       |
|-------------|-----------------------------|
| 2021-Sep-28 | Alignment with Azure style. |
| 2020-Oct-04 | Initial version in Wiki.    |

## Introduction

> When building a digital ecosystem you should use API-first approach and start with design and development of your APIs. Considering API usability and creating APIs that are easy to discover, simple to use, fit to purpose, and consistent across your products will make the

Reviewer: ```suggestion
When building a digital ecosystem you should use API-first approach and start with design and development of your APIs. Considering API usability and creating APIs that are easy to discover, simple to use, fit for purpose, and consistent across your products will make the
```

|---------------------------------------------------------------------------------------------|---------|
| :heavy_check_mark: **DO** return an error property with a child code property in all error responses.            | Error   |
| :heavy_check_mark: **DO** return a 403 Forbidden error when insufficient scopes are present on the auth token.   | Error   |
| :heavy_check_mark: **DO** return a 429 Too many requests error when the caller has exceeded throttling limits.   | Error   |
| :ballot_box_with_check: **CONSIDER** returning a 404 Not found error if a 403 would result in information disclosure. | Warning |

For a complete mapping of error codes to HTTP statuses you can refer to the
[rfc7231 (ietf.org)](https://datatracker.ietf.org/doc/html/rfc7231#section-6).

### API contract and non-backward compatible changes

Microsoft Graph definition of breaking changes is based on the [Microsoft REST
API
Guidelines](https://github.com/microsoft/api-guidelines/blob/graph/Guidelines.md#123-definition-of-a-breaking-change).
In general, making changes to the API contract for existing elements is
considered breaking. Adding new elements is allowed and not considered a
breaking change.

> :heavy_check_mark: **DO use not-breaking changes:**

Reviewer 2: ```suggestion
:heavy_check_mark: **DO use non-breaking changes:**
```
